server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -688,48 +688,59 @@ public AnalyzedRelation visitWithQuery(WithQuery node, StatementAnalysisContext
     protected AnalyzedRelation visitTable(Table<?> node, StatementAnalysisContext context) {
         QualifiedName tableQualifiedName = node.getName();
         SearchPath searchPath = context.sessionSettings().searchPath();
-        AnalyzedRelation relation;
         var relationContext = context.currentRelationContext();
 
         var withQuery = relationContext.parentSources()
             .getAncestor(RelationName.of(tableQualifiedName, null));
         if (withQuery != null) {
-            relation = withQuery;
+            relationContext.addSourceRelation(withQuery);
+            return withQuery;
         } else {
             TableInfo tableInfo;
-            try {
-                tableInfo = schemas.resolveTableInfo(
-                    tableQualifiedName,
-                    context.currentOperation(),
-                    context.sessionSettings().sessionUser(),
-                    searchPath
-                );
-                if (tableInfo instanceof DocTableInfo docTable) {
-                    // Dispatching of doc relations is based on the returned class of the schema information.
-                    relation = new DocTableRelation(docTable);
-                } else {
-                    relation = new TableRelation(tableInfo);
-                }
-            } catch (RelationUnknown e) {
-                View view;
+            var searchPathIterator = searchPath.iterator();
+            while (searchPathIterator.hasNext()) {
+                SearchPath currentSearchPath = SearchPath.createSearchPathFrom(searchPathIterator.next());
+                AnalyzedRelation relation;
                 try {
-                    view = schemas.resolveView(tableQualifiedName, searchPath);
-                } catch (RelationUnknown e1) {
-                    // don't shadow original exception, as looking for the view is just a fallback
-                    throw e;
+                    try {
+                        tableInfo = schemas.resolveTableInfo(
+                            tableQualifiedName,
+                            context.currentOperation(),
+                            context.sessionSettings().sessionUser(),
+                            currentSearchPath
+                        );
```

Comment:
`resolveTableInfo` is again iterating over the search path. Same for `resolveView`.
We could avoid that doing something like this:

```diff
diff --git a/libs/sql-parser/src/main/java/io/crate/sql/tree/QualifiedName.java b/libs/sql-parser/src/main/java/io/crate/sql/tree/QualifiedName.java
index 21c005add5b..c92a23e252d 100644
--- a/libs/sql-parser/src/main/java/io/crate/sql/tree/QualifiedName.java
+++ b/libs/sql-parser/src/main/java/io/crate/sql/tree/QualifiedName.java
@@ -28,10 +28,12 @@
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.StreamSupport;
 
+import org.jetbrains.annotations.Nullable;
+
 import io.crate.common.collections.Lists;
 
 public class QualifiedName {
 
     private final List<String> parts;
@@ -58,10 +60,24 @@ public QualifiedName(Iterable<String> parts) {
             throw new IllegalArgumentException("parts is empty");
         }
         this.parts = partsList;
     }
 
+    @Nullable
+    public String schemaName() {
+        if (parts.size() >= 2) {
+            return parts.get(parts.size() - 2);
+        }
+        return null;
+    }
+
+    public String tableName() {
+        assert parts.size() <=
+               3 : "When identifying schemas or tables a qualified name should not have more the 3 parts";
+        return parts.get(parts.size() - 1);
+    }
+
     public List<String> getParts() {
         return parts;
     }
 
     @Override
diff --git a/server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java b/server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
index bf6408ba4b5..49c86df810c 100644
--- a/server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
+++ b/server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java
@@ -694,33 +694,43 @@ protected AnalyzedRelation visitTable(Table<?> node, StatementAnalysisContext co
             .getAncestor(RelationName.of(tableQualifiedName, null));
         if (withQuery != null) {
             relationContext.addSourceRelation(withQuery);
             return withQuery;
         } else {
+
+            @Nullable
+            String schemaName = tableQualifiedName.schemaName();
+            String tableName = tableQualifiedName.tableName();
+
+            if (schemaName != null) {
+                RelationName relationName = new RelationName(schemaName, tableName);
+                try {
+                    TableInfo tableInfo = schemas.getTableInfo(relationName, context.currentOperation());
+                } catch (RelationUnknown e) {
+                    View view = schemas.getView(relationName);
+                }
+            }
+
             TableInfo tableInfo;
             var searchPathIterator = searchPath.iterator();
             while (searchPathIterator.hasNext()) {
-                SearchPath currentSearchPath = SearchPath.createSearchPathFrom(searchPathIterator.next());
+                String currentSchema = searchPathIterator.next();
+                RelationName relationName = new RelationName(currentSchema, tableName);
                 AnalyzedRelation relation;
                 try {
                     try {
-                        tableInfo = schemas.resolveTableInfo(
-                            tableQualifiedName,
-                            context.currentOperation(),
-                            context.sessionSettings().sessionUser(),
-                            currentSearchPath
-                        );
+                        tableInfo = schemas.getTableInfo(relationName, context.currentOperation());
                         if (tableInfo instanceof DocTableInfo docTable) {
                             // Dispatching of doc relations is based on the returned class of the schema information.
                             relation = new DocTableRelation(docTable);
                         } else {
                             relation = new TableRelation(tableInfo);
                         }
                     } catch (RelationUnknown e) {
                         View view;
                         try {
-                            view = schemas.resolveView(tableQualifiedName, currentSearchPath);
+                            view = schemas.getView(relationName);
                         } catch (RelationUnknown e1) {
                             // don't shadow original exception, as looking for the view is just a fallback
                             throw e;
                         }
                         ViewMetadata viewMetadata = view.metadata();
```

Reply:
Please see the [update](https://github.com/crate/crate/pull/15543#discussion_r1488761232).